torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -960,6 +1006,62 @@ void ProcessGroupNCCL::logWorkEnd(WorkNCCL& work) {
       store_, traceKeyEnd_, work.seq_, opTypeToString(work.opType_));
 }
 
+void ProcessGroupNCCL::workMonitorLoop() {
+  bool done = false;
+  // needs to use a copy since we must drop workMetaListMutex_ during the main
+  // loop
+  std::list<c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL>> metaCopy;
+  std::list<c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL>> cleanupCopy;
+  while (!done || !terminateProcessGroup_.load()) {
+    {
+      std::unique_lock<std::mutex> lock(workMetaListMutex_);
+      // We busy-poll the work vector every kWatchdogThreadSleepMillis
+      // milliseconds as long as the atomic is True.
+      workMetaListCV_.wait_for(
+          lock,
+          std::chrono::milliseconds(kWatchdogThreadSleepMillis),
+          [&]() -> bool { return terminateProcessGroup_.load(); });
+
+      lastMonitorStart_.store(std::chrono::steady_clock::now());
+      std::copy(
+          workMetaList_.begin(),
+          workMetaList_.end(),
+          std::back_inserter(metaCopy));
+      // we can't trigger the dtor here as we're holding a lock
+      std::copy(
+          workCleanupList_.begin(),
+          workCleanupList_.end(),
+          std::back_inserter(cleanupCopy));
+      workCleanupList_.clear();
+    }
+
+    // LOCKING: hanging hazzard - no locks allowed from here
```

Comment:
nit: comment should explain the rationale behind the 2 threads. IIUC is is

ncclMonitorThread: takes care of doing work cleanups.  but since cleanup may hang, monitorThread also may hang.  Hence, ensure that watchdogThread itself can never hang.  And therefore monitorThread must not hang with lock held, since lock is also needed by watchdog.